grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcHealthCheckService.java

```diff
@@ -0,0 +1,287 @@
+
+    private static final Logger logger = LoggerFactory.getLogger(GrpcHealthCheckService.class);
+
+    private static final HealthCheckResponse SERVING_RESPONSE =
+            HealthCheckResponse.newBuilder()
+                               .setStatus(ServingStatus.SERVING)
+                               .build();
+
+    private static final HealthCheckResponse NOT_SERVING_RESPONSE =
+            HealthCheckResponse.newBuilder()
+                               .setStatus(ServingStatus.NOT_SERVING)
+                               .build();
+
+    private static final HealthCheckResponse SERVICE_UNKNOWN_RESPONSE =
+            HealthCheckResponse.newBuilder()
+                               .setStatus(ServingStatus.SERVICE_UNKNOWN)
+                               .build();
+
+    /**
+     * Returns a newly created {@link GrpcHealthCheckService}
+     * with the specified {@link ListenableHealthChecker}s.
+     */
+    public static GrpcHealthCheckService of(ListenableHealthChecker... healthCheckers) {
+        return builder().checkers(healthCheckers).build();
+    }
+
+    /**
+     * Returns a newly created {@link GrpcHealthCheckService}
+     * with the specified {@link ListenableHealthChecker}s.
+     */
+    public static GrpcHealthCheckService of(Iterable<? extends ListenableHealthChecker> healthCheckers) {
+        return builder().checkers(healthCheckers).build();
+    }
+
+    /**
+     * Returns a new builder which builds a new {@link GrpcHealthCheckService}.
+     */
+    public static GrpcHealthCheckServiceBuilder builder() {
+        return new GrpcHealthCheckServiceBuilder();
+    }
+
+    private static HealthCheckResponse getHealthCheckResponse(ServingStatus status) {
+        if (status == ServingStatus.SERVING) {
+            return SERVING_RESPONSE;
+        } else if (status == ServingStatus.NOT_SERVING) {
+            return NOT_SERVING_RESPONSE;
+        } else if (status == ServingStatus.SERVICE_UNKNOWN) {
+            return SERVICE_UNKNOWN_RESPONSE;
+        } else {
+            throw new IllegalArgumentException("Invalid status:" + status);
+        }
+    }
+
+    private final SettableHealthChecker serverHealth;
+    private final Set<ListenableHealthChecker> healthCheckers;
+    private final Map<String, ListenableHealthChecker> grpcServiceHealthCheckers;
+    private final NonBlockingIdentityHashMap<StreamObserver<HealthCheckResponse>, Entry<String, ServingStatus>>
+            watchers = new NonBlockingIdentityHashMap<>();
+    @Nullable
+    private Server server;
+    private boolean updating;
+
+    GrpcHealthCheckService(
+            Set<ListenableHealthChecker> healthCheckers,
+            Map<String, ListenableHealthChecker> grpcServiceHealthCheckers,
+            List<HealthCheckUpdateListener> updateListeners
+    ) {
+        serverHealth = new SettableHealthChecker(false);
+        if (!updateListeners.isEmpty()) {
+            addServerHealthUpdateListener(ImmutableList.copyOf(updateListeners));
```

Comment:
```suggestion
            addServerHealthUpdateListener(updateListeners);
```
We just iterate the list in the method so we don't have to copy it.